openedx/core/lib/graph_traversals.py

```diff
@@ -0,0 +1,263 @@
+):
+    """
+    Generator for yielding nodes of a tree (or directed acyclic graph)
+    in a topological sort.  The tree is traversed using the
+    get_parents and get_children accessors.  The get_result function is
+    used to compute the yielded result value.  The filter_func function is
+    used to limit which nodes are actually yielded.
+
+    Arguments:
+        start_node (any hashable type) - The starting node for the
+        traversal.
+
+        get_parents (F(node)->[node]) - Function that returns a list of
+            parent nodes for a given node.
+
+        get_children (F(node)->[node]) - Function that returns a list of
+            children nodes for a given node.
+
+        get_result (F(node)->any type) - Function that computes and
+            returns the resulting value to be yielded for the given
+            node.  If None, the identity function is assumed.
+
+        filter_func (F(node)->boolean) - Function that returns
+            whether or not to yield the given node.
+            If None, the True function is assumed.
+
+        yield_descendants_of_unyielded (boolean) -
+            If False, descendants of an unyielded node are not
+                yielded.
+            If True, descendants of an unyielded node are yielded even
+                if none of their parents were yielded.
+
+            Note: In case of a DAG, a descendant is yielded if *any* of
+            its parents are yielded.
+
+    Yields:
+        get_result(node): The result of the next node in the traversal.
+    """
+    return _traverse_generic(
+        start_node,
+        get_parents=get_parents,
+        get_children=get_children,
+        get_result=get_result,
+        filter_func=filter_func,
+        yield_descendants_of_unyielded=yield_descendants_of_unyielded,
+    )
+
+
+def traverse_pre_order(start_node, get_children, get_result=None, filter_func=None):
+    """
+    Generator for yielding nodes of a tree (or directed acyclic graph)
+    in a pre-order sort.
+
+    Arguments:
+        See description in traverse_topologically.
+    """
+    return _traverse_generic(
+        start_node,
+        # There's no need to get_parents
+        get_parents=None,
+        get_children=get_children,
+        get_result=get_result,
+        filter_func=filter_func,
+    )
+
+
+def traverse_post_order(start_node, get_children, get_result=None, filter_func=None):
+    """
+    Generator for yielding nodes of a tree (or directed acyclic graph)
+    in a post-order sort.
+
+    Arguments:
+        See description in traverse_topologically.
+    """
+    class _Node(object):
+        """
+        Wrapper class for nodes that keeps a children iterator.
+        """
+        def __init__(self, node, get_children):
+            self.node = node
+            self.children = iter(get_children(node))
+
+    # If get_result or filter_func aren't provided, make them no-ops.
+    get_result = get_result or (lambda node_: node_)
```

Comment:
generally when using a variable name that conflicts with a system name.
regardless, this line is now removed.